lua/lean/infoview.lua

```diff
@@ -61,42 +69,51 @@ function Infoview:new(id, width)
   -- Make sure we notice even if someone manually :q's the infoview window.
   set_augroup("LeanInfoviewClose", string.format([[
     autocmd WinClosed <buffer> lua require'lean.infoview'.__was_closed(%d)
-  ]], id))
-
-  local obj = { bufnr = bufnr, window = window }
-  setmetatable(obj, self)
-  self.__index = self
+  ]], self.id))
 
   vim.api.nvim_set_current_win(window_before_split)
 
-  obj.focus_on_current_buffer()
+  self.bufnr = bufnr
+  self.window = window
+  self.is_open = true
 
-  return obj
-end
+  self:focus_on_current_buffer()
 
---- Do nothing, we're already open.
-function Infoview:open()
-  return self
+  return vim.deepcopy(self)
 end
 
 --- Close this infoview.
-function Infoview:close()
-  vim.api.nvim_win_close(self.window, true)
-  set_augroup("LeanInfoviewUpdate", "", true)
+function Infoview:close(win_already_closed)
+  if not self.is_open then
+    -- in case it is nil
+    self.is_open = false
+    return
+  end
+
+  if not win_already_closed then vim.api.nvim_win_close(self.window, true) end
+  self.bufnr = nil
+  self.window = nil
+  self.is_open = false
+
+  self:focus_on_current_buffer()
 end
 
 --- Toggle this infoview being open.
 function Infoview:toggle()
-  self:close()
+  if self.is_open then self:close() else self:open() end
```

Comment:
> Consider removing Infoview/ClosedInfoview.is_open or replacing with a call to vim.fn.win_findbuf or filtering on window names? Right now only tests access this attribute (essentially to differentiate between the two), but it'd seem better if those tests instead actually look to see whether there's an infoview window visible.

We're now using this field here, as well as a couple of other places. `vim.fn.win_findbuf` is a good idea as an additional testing criterion, but I think it wouldn't hurt to leave in the `is_open` check just as a quick internal sanity check for us.

Reply:
Sounds good to me!